Allow "add address" and "add customer" and "edit address" buttons to be handled in modal

matks opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Allow "add address" and "add customer" buttons to be handled in modal
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |  https://github.com/PrestaShop/PrestaShop/issues/10583
| How to test?  | Previously, in migrated page "Create an order" ; "add address" and "add customer" were links, so clicking them would make you exit the page. Now they open modals so you dont exit the page.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/17090)
<!-- Reviewable:end -->

matks committed:
Allow Create an Order to use a modal for Create a Customer button

matks committed:
Allow Create an order 'add an address' button to be handled inside a modal

Progi1984 commented on admin-dev/themes/new-theme/js/pages/order/create.js:
```diff
@@ -25,7 +25,28 @@
+ */
+function searchCustomerByString(string)
+{
+  orderPageManager.customerManager.search(string);
```
May be a test for checking `orderPageManager` is not null.

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js:
```diff
@@ -65,6 +65,10 @@ export default class CreateOrderPage {
     this._loadCartFromUrlParams();
+
+    return {
+      refreshAddressesList: () => this._refreshAddressesList(),
```
```suggestion
      refreshAddressesList: () => this.refreshAddressesList(),
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js:
```diff
@@ -128,6 +132,18 @@ export default class CreateOrderPage {
     this.$container.on('blur', createOrderMap.cartRuleSearchInput, () => this.cartRuleManager.stopSearching());
     this._listenForCartEdit();
     this._onCartLoaded();
+    this._initAddAddressIframe();
```
```suggestion
    this.initAddAddressIframe();
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js:
```diff
@@ -128,6 +132,18 @@ export default class CreateOrderPage {
+  /**
+   * @private
+   */
+  _initAddAddressIframe() {
```
```suggestion
  initAddAddressIframe() {
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js:
```diff
@@ -145,27 +161,27 @@ export default class CreateOrderPage {
 
     this.$container.on('change', createOrderMap.deliveryOptionSelect, e =>
-      this.cartEditor.changeDeliveryOption(this.cartId, e.currentTarget.value),
+        this.cartEditor.changeDeliveryOption(this.cartId, e.currentTarget.value),
```
2 spaces for indentation in javascript files :)

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js:
```diff
@@ -470,4 +487,16 @@ export default class CreateOrderPage {
+   *
+   * @private
+   */
+  _refreshAddressesList() {
```
```suggestion
  refreshAddressesList() {
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js:
```diff
@@ -470,4 +487,16 @@ export default class CreateOrderPage {
+   */
+  _refreshAddressesList() {
+    const cartId = $(createOrderMap.cartBlock).data('cartId');
+    $.get(this.router.generate('admin_carts_info', {cartId})).then((cartInfo) => {
```
No catch?

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/customer-manager.js:
```diff
@@ -46,6 +46,7 @@ export default class CustomerManager {
     this.customerRenderer = new CustomerRenderer();
 
     this._initListeners();
+    this._initAddCustomerIframe();
```
```suggestion
    this.initAddCustomerIframe();
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/customer-manager.js:
```diff
@@ -66,6 +67,17 @@ export default class CustomerManager {
+  /**
+   * @private
+   */
+  _initAddCustomerIframe() {
```
```suggestion
  initAddCustomerIframe() {
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/order/create/customer-manager.js:
```diff
@@ -88,6 +100,15 @@ export default class CustomerManager {
+            'liteDisplaying': 1,
+            'submitFormAjax': 1,
+            'id_customer': this.customerId,
+          });
```
```suggestion
          },
        );
```

PierreRambaud commented on src/PrestaShopBundle/Controller/Admin/Sell/Address/AddressController.php:
```diff
@@ -307,6 +307,10 @@ public function createAction(Request $request): Response
             if ($handlerResult->isSubmitted() && $handlerResult->isValid()) {
                 $this->addFlash('success', $this->trans('Successful creation.', 'Admin.Notifications.Success'));
 
+                if ($request->query->has('submitFormAjax')) {
```
Why not using `$request->isXmlHttpRequest()`?

matks replied:
It would be better, but I cannot guarantee all legacy calls using `submitFormAjax` parameter are proper `proper http request`.

In order to comply 100% with legacy behavior and ensure backward compatibility, I stay with previous implementation rules.

PierreRambaud replied:
:sob:

matks replied:
I can add a `@todo PS 8.0.0` if you want 😅

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Address/modal_create_success.html.twig:
```diff
@@ -0,0 +1,47 @@
+    {# Script dedicated to close "Create an address" fancybox used in Add an Order #}
+
+    // legacy behavior
+    if (typeof parent.getSummary !== "undefined") {
```
```suggestion
    if (typeof window.parent.getSummary !== "undefined") {
```

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Address/modal_create_success.html.twig:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    // symfony behavior
+    if (typeof parent.order_create !== "undefined") {
```
```suggestion
    if (typeof window.parent.order_create !== "undefined") {
```

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Customer/modal_create_success.html.twig:
```diff
@@ -31,8 +31,18 @@
-    parent.$.fancybox.close();
+
+    // legacy behavior
+    if (typeof parent.setupCustomer !== "undefined") {
```
```suggestion
    if (typeof window.parent.setupCustomer !== "undefined") {
```

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Customer/modal_create_success.html.twig:
```diff
@@ -31,8 +31,18 @@
+      window.parent.setupCustomer({{ customerId }});
+    }
+    // symfony behavior
+    if (typeof parent.order_create !== "undefined") {
```
```suggestion
    if (typeof window.parent.order_create !== "undefined") {
```

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Customer/modal_create_success.html.twig:
```diff
@@ -31,8 +31,18 @@
+    }
+    // symfony behavior
+    if (typeof parent.order_create !== "undefined") {
+      $('#customer-search-input', window.parent.document).val('{{ customerEmail }}');
```
You should escape customerEmail for js
```suggestion
      $('#customer-search-input', window.parent.document).val('{{ customerEmail|e('js') }}');
```

matks replied:
Nice one 👍

matks committed:
Handle code review feedbacks

matks commented:
@PierreRambaud your feedbacks have been handled into commit https://github.com/PrestaShop/PrestaShop/pull/17090/commits/be2fd7530a7110fa245c0049597cb3121383fde7

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Customer/modal_create_success.html.twig:
```diff
@@ -31,8 +31,18 @@
+
+    // legacy behavior
+    if (typeof window.parent.setupCustomer !== "undefined") {
+      $('#customer', window.parent.document).val('{{ customerEmail }}');
```
the `customerEmail` must be escape everywhere :D same for `customerId`

matks commented:
I dont know why but I checked out this branch this morning and I get a `fancybox is not a function` TypeError 😅

PierreRambaud commented:
> I dont know why but I checked out this branch this morning and I get a `fancybox is not a function` TypeError

TROLOLOLO

matks committed:
Handle code review feedbacks - part 2

matks committed:
Rebuild Create an Order assets

matks commented:
@PierreRambaud Fixed and rebased

matks committed:
Build assets

matks commented:
@Robin-Fischer-PS rebased ✅

Robin-Fischer-PS commented:
Hi @matks !

Some bugs and a question : 

-When canceling the add of an address, instead of closing the popin, the popin stays displayed and redirect to address page
![pr17090 cancel add address](https://user-images.githubusercontent.com/55688534/73258535-2fd46480-41c6-11ea-8842-e2e5d87faed3.png)
- After creating the first address of a customer, the address is not displayed
![pr17090 blank block ](https://user-images.githubusercontent.com/55688534/73258483-13d0c300-41c6-11ea-92b5-d2f1c061504b.png)
- Edit address should be in modale too, so is "Details" of a customer. Will this be done in another PR ?

Thanks !

PierreRambaud commented:
@Robin-Fischer-PS It's not a bug, it's a feature!

matks committed:
Handle fancybox close for Customer and Addresses create form

matks committed:
Fix some exotic behaviors and allow Edit Customer btn to be handled by fancybox in Create Order page

matks commented:
> Hi @matks !
> 
> Some bugs and a question :
> 
> -When canceling the add of an address, instead of closing the popin, the popin stays displayed and redirect to address page

Fixed

> * After creating the first address of a customer, the address is not displayed

Fixed

> * Edit address should be in modale too, so is "Details" of a customer. Will this be done in another PR ?

Fixed

Please note that there is a difference in behavior for "create address button":
- if, when you click it, there is no address linked to customer, then after successfull address creation I automatically link this new address to the cart being created in the page (which fixes _After creating the first address of a customer, the address is not displayed_)
- if, when you click it, there are already some addresses available for the customer, I only add the new address into the dropdown, I dont automatically assign it

Robin-Fischer-PS commented:
Hi @matks 

One bug that I should have seen before (sorry) : 
- When I add an address that is not usable for this order (for example, an address in a country not accepted by the shop), there is an unclear error and the address is not added. In Legacy, the address is added, but there is no carrier shown, blocking the creation of the order. I supposed this is caused by the fact that, on Migrated Create an order page, there is only the usable addresses that are listed, in Legacy, all addresses are listed.
![pr17090 invalid address](https://user-images.githubusercontent.com/55688534/73443950-ed429180-4357-11ea-9ff5-cbefdc49b983.png)

matks commented:
> * When I add an address that is not usable for this order (for example, an address in a country not accepted by the shop), there is an unclear error and the address is not added.

I cant reproduce the error.

However indeed the address is not added to dropdown if it is not suitable for this order or the shop settings.
Should I display an error message "Address added successfully however it cannot be used for this order as it does meet the requirements of this cart" ?

It seems a bit over-engineered to me 🤔 but if you tell me to do it, I'll do it

matks commented:
(adding "waiting for QA" label to get an answer)

Robin-Fischer-PS commented:
Hi @matks ,

If the "weird" error does not appear anymore, there is indeed no need to display another message. However, since I can't install the PR because of conflict, I cannot try to reproduce it.

matks commented:
PR rebased, ready to QA when I get the 2 approvals

matks committed:
Rebuild assets

NeOMakinG commented:
Can you please resolve the conflict @matks ? Thanks :)

matks committed:
Rebase and rebuild assets

matks commented:
Rebased

Robin-Fischer-PS commented:
Hi @matks !

Since I can't remember what was said on the following topic, I ping you again : 
"Edit address should be in modale too, so is "Details" of a customer. Will this be done in another PR ?"

Otherwise, this PR is OK for me, so as soon as you answer me, I will be able to QA-approve it :yum:

Robin-Fischer-PS commented:
As seen with @matks , this will be done in another PR.

Issue is here : https://github.com/PrestaShop/PrestaShop/issues/17223 (point 3).

Progi1984 commented:
Hello @matks, thanks for your contribution, would you mind to rebase your pull request? Thanks

matks commented:
> Hello @matks, thanks for your contribution, would you mind to rebase your pull request? Thanks

<img src="https://i.redd.it/93iuffshbit21.jpg">

matks committed:
Rebuild assets
